### PR TITLE
Add print_version and replace_version to module.py 

### PR DESCRIPTION
**Pull request by balopat**
Adds print_version and replace_version to dev_tools/modules.py.

This helps with #4129 - automating releases.

### CODE REVIEW

**Comment by balopat** [comment]
@MichaelBroughton - in preparation for our v0.12 release.

File: dev_tools/modules.py [hunk 1]
```diff
@@ -101,6 +113,66 @@ def list_modules(
...
+
+    for m in list_modules(search_dir=search_dir, include_parent=True):
+        version_file = _find_version_file(search_dir / m.root)
+        content = version_file.read_text("UTF-8")
+        new_content = content.replace(old, new)
+        version_file.write_text(new_content)
+
+
+def _validate_version(new_version: str):
+    if not re.match(r"\d+\.\d+\.\d+(\.dev)?", new_version):
```
Comment (MichaelBroughton) [inline dev_tools/modules.py:165]: Consider using regex from here: https://semver.org/

**Review by MichaelBroughton — APPROVED** [review]
LGTM

File: dev_tools/modules.py (hunk 1 above)
Comment (balopat) [reply dev_tools/modules.py:165]: Oh, cool, didn't know about it - however, interestingly it doesn't seem to match `1.2.3.dev`

Comment (balopat) [reply dev_tools/modules.py:165]: It does match `1.2.3-dev`...hmmm...

**Pull request merged**